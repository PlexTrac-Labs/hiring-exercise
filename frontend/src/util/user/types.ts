
export type CreateUserRequest = {
    username: string;
    password: string;
    firstName: string;
    lastName: string;
    email: string;
    birthYear: number;
    favColor: string;
}

export type UpdateUserRequest = {
    username: string;
    firstName?: string;
    lastName?: string;
    email?: string;
    birthYear?: number;
    favColor?: string;
};

export type User = {
    _id: string;
    username: string;
    password: string;
    firstName: string;
    lastName: string;
    email: string;
    birthYear: number;
    favColor: string;
    admin: boolean;
}

export type AuthResponse = {
    auth_token: string;
    user: User;
}

export type ResetPasswordRequest = {
    oldPassword: string;
    newPassword: string;
}
