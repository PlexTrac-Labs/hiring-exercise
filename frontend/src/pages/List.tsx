import { Typography } from "@mui/material";
import React from "react";

export const List = () => {
  return <Typography>List</Typography>;
};
