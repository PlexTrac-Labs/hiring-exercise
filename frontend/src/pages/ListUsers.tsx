import { Grid, List, Typography } from "@mui/material";
import React, { useEffect, useState } from "react";
import { CreateUser } from "../components/CreateUser";
import { EditDialog } from "../components/EditDialog";
import { UserListItem } from "../components/UserListItem";
import { User } from "../util/user/types";
import { deleteUser, getAllUsers } from "../util/user/user";

export const ListUsers = () => {
  const [openEditDialog, setOpenEditDialog] = useState(false);
  const [users, setUsers] = useState<User[]>([]);
  const [selectedUser, setSelectedUser] = useState<User>({
    _id: "",
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: "",
    birthYear: 0,
    favColor: "",
    admin: false,
  });

  const handleDelete = (id: string, index: number) => {
    deleteUser(id);
    users.splice(index, 1);
    setUsers([...users]);
  };

  const handleUserClick = (user: User) => {
    setSelectedUser(user);
    setOpenEditDialog(true);
  };

  useEffect(() => {
    let getAll = async () => {
      let response = await getAllUsers();
      setUsers(response);
    };
    getAll();
  }, []);

  return (
    <Grid
      data-testid="listUsers"
      container
      direction="column"
      rowSpacing={2}
      alignItems="center"
      justifyContent="center"
    >
      <Grid item xs={12}>
        <Typography variant="h3">Users</Typography>
        <EditDialog
          user={selectedUser}
          open={openEditDialog}
          setUsers={setUsers}
          onClose={() => setOpenEditDialog(false)}
        />
      </Grid>
      <Grid item xs={12}>
        <List>
          <CreateUser setUsers={setUsers} />
          {users.map((user, index) => (
            <UserListItem
              key={user._id}
              user={user}
              handleClick={handleUserClick}
              handleDelete={() => handleDelete(user._id, index)}
            />
          ))}
        </List>
      </Grid>
    </Grid>
  );
};
