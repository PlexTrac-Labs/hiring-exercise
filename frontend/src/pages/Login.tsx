import {
  Alert,
  Box,
  Button,
  FormControl,
  Grid,
  TextField,
  Typography,
} from "@mui/material";
import { styled } from "@mui/system";
import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { AUTH_TOKEN } from "../helpful/constants";
import { authenticate } from "../util/user/user";

const LoginBox = styled(Box)(({ theme }) => ({
  display: "flex",
  flexDirection: "column",
  padding: 15,
  //backgroundColor: "gray",
  justifyContent: "center",
  alignItems: "center",
  //height: "100vh",
}));

export const Login = () => {
  const navigate = useNavigate();
  const [error, setError] = useState(false);

  const handleSubmit = async (event: any) => {
    event.preventDefault();
    const username = event.target.username.value;
    const password = event.target.password.value;
    try {
      const auth = await authenticate(username, password);
      localStorage.setItem(AUTH_TOKEN, auth.auth_token);
      navigate("list");
    } catch (error) {
      setError(true);
    }
  };

  return (
    <LoginBox>
      <form onSubmit={handleSubmit}>
        <Grid
          container
          direction="column"
          rowSpacing={2}
          alignItems="center"
          justifyContent="center"
        >
          <Grid item xs={12}>
            <Typography variant="h3">Login</Typography>
          </Grid>
          {error && <Alert severity="error">Invalid Username/Password</Alert>}

          <Grid item xs={12}>
            <FormControl>
              <TextField
                required
                id="username"
                variant="outlined"
                label="Username"
              />
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <FormControl>
              <TextField
                required
                id="password"
                variant="outlined"
                label="Password"
                type="password"
              />
            </FormControl>
          </Grid>
          <Grid item xs={12}>
            <Button color="primary" variant="contained" type="submit">
              Login
            </Button>
          </Grid>
        </Grid>
      </form>
    </LoginBox>
  );
};
