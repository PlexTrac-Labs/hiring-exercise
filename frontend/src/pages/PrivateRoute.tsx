import React from "react";
import { Navigate, Outlet } from "react-router-dom";
import { AUTH_TOKEN } from "../helpful/constants";

export const PrivateRoute = () => {
  const loggedIn = !!localStorage.getItem(AUTH_TOKEN);

  return loggedIn ? <Outlet /> : <Navigate to="/" />;
};
