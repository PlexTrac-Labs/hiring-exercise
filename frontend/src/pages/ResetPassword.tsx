import { Grid, Typography } from "@mui/material";
import React from "react";

export const ResetPassword = () => {
  return (
    <Grid
      container
      direction="column"
      rowSpacing={2}
      alignItems="center"
      justifyContent="center"
    >
      <Grid item xs={12}>
        <Typography variant="h3">Reset Password</Typography>
      </Grid>
    </Grid>
  );
};
