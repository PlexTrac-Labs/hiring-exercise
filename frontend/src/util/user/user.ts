import axios from "axios";
import { CreateUserRequest, UpdateUserRequest } from "./types"


// POST
export const createUser = (payload: CreateUserRequest) => {
    console.log('kdp createUser');
}

export const authenticate = (username: string, password: string) => { //kdp revisit these params
    console.log('kdp authenticate');
    return axios.post(`http://localhost:5000/authenticate`, { username, password})
        .then(res => res.data);
}

// PUT
export const updateUser = (payload: UpdateUserRequest) => {
    console.log('kdp update user')
}

//DELETE
export const deleteUser = (id: string) => {
    console.log('kdp deleteUser')
}

// GET

export const getUser = (id: string) => {
    return axios.get(`http://localhost:5000/user/${id}`)
        .then(res => res.data)
}

export const getAllUsers = () => {
    return axios.get(`http://localhost:5000/user`)
    .then(res => res.data)
}