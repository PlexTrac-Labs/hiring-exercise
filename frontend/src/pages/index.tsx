import React from "react";
import { List } from "./List";
import { Login } from "./Login";
import { User } from "./User";

export type PageType = {
  title: string;
  route: string;
  comp: React.FunctionComponent;
};

const Pages: PageType[] = [
  {
    title: "Login",
    route: "/",
    comp: Login,
  },
  {
    title: "List",
    route: "/list",
    comp: List,
  },
  {
    title: "User",
    route: "/user",
    comp: User,
  },
];

export default Pages;
