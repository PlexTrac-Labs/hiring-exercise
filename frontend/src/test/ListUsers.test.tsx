import { cleanup, render, waitFor } from "@testing-library/react";
import React from "react";
import { ListUsers } from "../pages/ListUsers";
import { getAllUsers } from "../util/user/user";
import { EMPTY_USER } from "./testConstants";

jest.mock("../util/user/user", () => ({
  getAllUsers: jest.fn(),
}));

const mockedGetAllUsers = getAllUsers as jest.Mock;

afterEach(cleanup);

it("renders", async () => {
  mockedGetAllUsers.mockImplementation(() => [EMPTY_USER]);
  const { getByTestId, asFragment } = render(<ListUsers />);
  await waitFor(() => getByTestId("listUsers"));
  expect(asFragment()).toMatchSnapshot();
});
