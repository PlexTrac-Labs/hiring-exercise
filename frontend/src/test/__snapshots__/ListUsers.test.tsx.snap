// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined MuiGrid-direction-xs-column css-1yhrxth-MuiGrid-root"
    data-testid="listUsers"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    >
      <h3
        class="MuiTypography-root MuiTypography-h3 css-gepadz-MuiTypography-root"
      >
        Users
      </h3>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1em9ut6-MuiGrid-root"
    >
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-16a21k1-MuiListItem-root"
        >
          <div
            class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium css-4b67si-MuiSvgIcon-root"
                data-testid="AddCircleIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
                />
              </svg>
            </div>
            Create User
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
        </li>
        <li
          class="MuiListItem-root MuiListItem-gutters MuiListItem-padding css-f73flm-MuiListItem-root"
        >
          <div
            class="MuiListItemButton-root MuiListItemButton-gutters MuiButtonBase-root css-cvhtoe-MuiButtonBase-root-MuiListItemButton-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              />
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
          <div
            class="MuiListItemSecondaryAction-root css-518kzi-MuiListItemSecondaryAction-root"
          >
            <button
              aria-label="Delete User"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
              data-mui-internal-clone-element="true"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
